[PLATFORM-286] Refactor withWeb3 HOC

juhah opened the pull request:
We need to use a modal in profile settings to connect an ethereum identity and with that we need all the usual checks for Metamask. I refactored the `withWeb3` HOC into a shared utility and the web3 validation is now done so that it just throws errors if anything goes wrong. That way it can be potentially used outside of the dialogs as well. I also moved the account info polling from Marketplace's `GlobalInfoWatcher` to a shared class so it can be used by the HOC. The polling is needed to check the metamask account status when the modal is open, and it uses an `EventEmitter` that you can subscribe to.

There are a lot of changes due to moved files but main changes are the `Web3Poller` class, the `withWeb3` HOC and the `validateWeb3` utility. Next step would be the refactor the `withContractProduct` HOC but I left it as it is for now.

Summary of changes:
- moved png images to shared, created new `PngImage` component
- moved web3 account polling to a separate class called `Web3Poller`
- Refactored `GlobalInfoWatcher` to subscribe to `Web3Poller`
- Added a utility called `validateWeb3` that throws errors if not supported. It will also prompt permission from Metamask if logged out. 
- Moved some web3 utils to shared from marketplace
- Moved some error dialogs to shared

Can be tested in profile settings by adding a new ethereum address and in marketplace (purchase, publish, save). Try logging out and then opening the dialog and then logging out while the dialog is open, try declining Metamask permission, try a non supported browser, etc. 

I'll update some more tests still to this branch but I wanted to push the code for review already.

juhah commented on app/src/shared/utils/withWeb3.jsx:
```diff
@@ -0,0 +1,102 @@
+            })
+        }
+
+        validateWeb3 = () => {
```
That's a very good idea, actually. It should play nicely with checking the contract product as well we refactor the `withContractProduct` HOC. Maybe we don't even need the HOC at all?

mondoreale replied:
It'd be great if that's possible, totally!

timoxley commented on app/src/shared/utils/withWeb3.jsx:
```diff
@@ -0,0 +1,102 @@
+                web3Error: null,
+            }, () => {
+                validateWeb3(getWeb3())
+                    .then(() => {
```
when using async + setState always need to check whether component has unmounted, otherwise React prints annoying "trying to set state on unmounted component" console warning

timoxley commented on app/src/shared/utils/withWeb3.jsx:
```diff
@@ -0,0 +1,102 @@
+        }
+
+        state = {
+            checkingWeb3: false,
```
perhaps:
```js
checkingWeb3: this.props.requireWeb3
```
This skips the first render with `checkingWeb3: false` immediately re-rendering with `checkingWeb3: true`

juhah replied:
It needs to be `false` to start the initial validation but I could add a flag to the `validateWeb3` method.

timoxley commented on app/src/shared/web3/web3Poller.js:
```diff
@@ -0,0 +1,192 @@
+    clearWeb3Poll = () => {
+        if (this.web3PollTimeout) {
+            clearTimeout(this.web3PollTimeout)
+            this.web3PollTimeout = undefined
```
@mondoreale i don't think that's right, clearing shouldn't mutate the value.

timoxley commented on app/src/shared/web3/web3Poller.js:
```diff
@@ -0,0 +1,192 @@
+    }
+
+    fetchWeb3Account = () => {
+        this.web3.getDefaultAccount()
```
should probably return the promise so other stuff can potentially chain around it or catch errors i.e. `return this.web3.getDefaultAccount()`

juhah commented on app/src/userpages/components/ProfilePage/IdentityHandler/AddIdentityButton/index.jsx:
```diff
@@ -0,0 +1,61 @@
+        })
+    }
+
+    onSave = () => {}
```
Yea, the actual feature of using modals to add an identity is actually missing - I'll implement that in another branch.

timoxley commented on app/src/shared/web3/web3Poller.js:
```diff
@@ -0,0 +1,192 @@
+                // if any other web3 actions are dispatched.
+                return Promise.resolve()
+            }, (err) => {
+                if (this.account !== null) {
```
perhaps should check for any falsey value i.e. `if (this.account)`

timoxley commented on app/src/shared/web3/web3Poller.js:
```diff
@@ -0,0 +1,192 @@
+                this.handleNetwork(network.toString())
+            }, () => {
+                this.web3.getDefaultAccount()
+                    .then(() => Promise.resolve(), (err) => {
```
I mean, why add `.then` at all if you're noop-ing it, just use `.catch`, or is this some weirdness enforced by the web3 lib?

juhah replied:
I think this was left over from before when we were dispatching some other things with it and not returning a resolved promise caused a warning about creating promise inside a handler. Removed it now.

timoxley commented on app/src/shared/web3/web3Poller.js:
```diff
@@ -0,0 +1,192 @@
+            .forEach((txHash) => {
+                // Get current number of confirmations and compare it with sought-for value
+                hasTransactionCompleted(txHash)
+                    .then((completed) => {
```
Lots of uncatchable async going on in here. Whenever you create a promise, you should always return it otherwise nothing can possible handle any errors that occur except the global event handler.

i.e. this should probably be something like:
```js
return Promise.all(Object.keys(getTransactionsFromSessionStorage()).map((txHash) => {
  return hasTransactionCompleted(txHash)
    .then((completed) => {
      if (completed) {
        return web3.eth.getTransactionReceipt(txHash) // always return created promises
          .then((receipt) => {
             if (receipt) {
               // …
             }
          })
    })
}))

```

timoxley replied:
or using async/await
```js
return Promise.all(Object.keys(getTransactionsFromSessionStorage()).map(async (txHash) => {
    const completed = await hasTransactionCompleted(txHash)
    if (completed) {
        const receipt = await web3.eth.getTransactionReceipt(txHash)
        if (receipt) {
            // …
        }
    }
})
```

juhah replied:
This is a lot cleaner, I used the latter option :)

timoxley commented on app/src/shared/web3/web3Poller.js:
```diff
@@ -0,0 +1,192 @@
+    }
+
+    fetchChosenEthereumNetwork = () => {
+        this.web3.getEthereumNetwork()
```
should always return promises

timoxley commented on app/src/shared/web3/web3Poller.js:
```diff
@@ -0,0 +1,192 @@
+    pollWeb3 = () => {
+        this.fetchWeb3Account()
+        this.clearWeb3Poll()
+        this.web3PollTimeout = setTimeout(this.pollWeb3, ONE_SECOND)
```
What happens if `fetchWeb3Account` takes longer than one second for whatever reason? it looks like it'll continue to enqueue more `fetchWeb3Account` calls every second without waiting for the previous one to resolve. If you return a promise from `fetchWeb3Account` then you could chain the `setTimeout` to occur only after `fetchWeb3Account` finishes, and perhaps should add a check to ensure it's not already running before starting a new one.

juhah replied:
Good point, changed it so that the new timeout is set after the promise completes. The old one is cleared in eg. `this.clearWeb3Poll` so basically we shouldn't end up in a situation where a timer is already set?

timoxley commented on app/src/shared/web3/web3Provider.js:
```diff
@@ -41,9 +49,45 @@ export const getWeb3 = (): StreamrWeb3 => {
+            throw new WalletLockedError()
+        }
+    } catch (e) {
+        throw new WalletLockedError()
```
Should probably at least log the error as a warning so it's easier to debug if something unexpected happens inside `await _web3.eth.getAccounts()`

timoxley commented on app/src/userpages/components/ProfilePage/IdentityHandler/AddIdentityButton/index.jsx:
```diff
@@ -0,0 +1,61 @@
+                >
+                    <Translate value="userpages.profilePage.ethereumAddress.addNewAddress" />
+                </Button>
+                {modalOpen && (
```
`!!modalOpen` for good measure.

mondoreale commented on app/src/shared/errors/Web3/index.jsx:
```diff
@@ -0,0 +1,57 @@
+    constructor(message: string = 'Web3 not supported') {
+        super(message)
+
+        // This is because of some bug in babel
```
Looks like it's this [bug](https://github.com/babel/babel/issues/4485) – extending `Error` doesn't work properly out of the box. Maybe we could use [`babel-plugin-transform-builtin-extend`](https://github.com/loganfsmyth/babel-plugin-transform-builtin-extend)?

juhah replied:
That seems like it, the comments were left over from me copy-pasting code from `TransactionError`

juhah commented on app/src/shared/utils/web3.js:
```diff
@@ -0,0 +1,24 @@
+        const { networkId: requiredNetwork } = getConfig()
+        const requiredNetworkName = ethereumNetworks[requiredNetwork]
+        const currentNetworkName = ethereumNetworks[network] || `#${network}`
+        if (network.toString() !== requiredNetwork.toString()) {
```
I'm not actually sure why the string comparison (or the `#{number}` fallback) is actually necessary, the current api documentation says that `web3.eth.net.getId()` will return a `Number`... I'll have to look into it why it was implemented like this before.

juhah commented on app/src/shared/components/Web3ErrorDialog/GenericErrorDialog/index.jsx:
```diff
@@ -0,0 +1,34 @@
+}: Props) => (
+    <Dialog
+        onClose={onClose}
+        title={!waiting ? title || I18n.t('modal.unlockWallet.title') : I18n.t('modal.genericError.waiting')}
```
True, maybe best to rename this to something like `UnlockWalletDialog`.

juhah commented on app/src/shared/utils/withWeb3.jsx:
```diff
@@ -0,0 +1,102 @@
+        }
+
+        componentWillUnmount() {
+            if (this.props.requireWeb3) {
```
Thanks, very good point!

juhah commented:
Thanks for the good feedback @mondoreale and @timoxley! I updated some of the changes, maybe I'll tackle the `web3` error boundary in another issue since this is already quite big.

timoxley commented on app/src/shared/web3/web3Poller/web3Poller.js:
```diff
@@ -0,0 +1,185 @@
+                    }
+                }
+            }
+        }))
```
you may also want to emit transaction `TRANSACTION_ERROR` if any of the async calls in this function happen to error e.g. `getTransactionReceipt`, or it might be fine for such an error to bubble up/crash? i.e. `try/catch` around the entire function with a `this.emitter.emit(events.TRANSACTION_ERROR, …)` in the catch, though this will emit a new error if any of the event handlers throw… so consider the following:
```js
    return Promise.all(Object.keys(getTransactionsFromSessionStorage()).map(async (txHash) => {
        let completed
        let receipt
        try {
            completed = await hasTransactionCompleted(txHash)
            receipt = !!completed && await web3.eth.getTransactionReceipt(txHash)
        } catch (err) {
            console.warn(err) // log unexpected errors
            this.emitter.emit(
                events.TRANSACTION_ERROR,
                txHash,
                new TransactionError(I18n.t('error.txFailed')), // could perhaps just pass the caught err or maybe TransactionError should know how to wrap another error?
            )
            return // bail out
        }
        if (completed && receipt) {
            // Cannot trust that receipt won't be null... the next interval should receive it
            if (receipt.status === true) {
                this.emitter.emit(
                    events.TRANSACTION_COMPLETE,
                    txHash,
                    receipt,
                )
            } else {
                this.emitter.emit(
                    events.TRANSACTION_ERROR,
                    txHash,
                    new TransactionError(I18n.t('error.txFailed'), receipt),
                )
            }
        }
    }))
```

juhah replied:
The error in this case is a bit misleading, the emitted error should reflect the transaction failing or succeeding in the blockchain. But I've added the logging at least if something happens here.

juhah commented on app/src/shared/web3/web3Poller/web3Poller.js:
```diff
@@ -0,0 +1,185 @@
+                if (this.account) {
+                    this.account = null
+                    this.emitter.emit(events.ACCOUNT_ERROR, err)
+                }
```
Failing here is ok since this method is called every second and the error is emitted only when a user was signed in already. Otherwise we will be spamming the error to console.

juhah commented on app/src/shared/web3/web3Provider.js:
```diff
@@ -4,15 +4,28 @@ import Web3 from 'web3'
 export class StreamrWeb3 extends Web3 {
+    isLegacy: boolean
+
+    constructor(provider: any, isLegacy: boolean = false) {
```
Good idea, took it into use!

juhah commented on app/src/shared/web3/web3Poller/web3Poller.js:
```diff
@@ -0,0 +1,185 @@
+                    .catch((err) => {
+                        if (this.account !== null) {
+                            this.emitter.emit(events.NETWORK_ERROR, err)
+                        }
```
The network polling is actually useless atm since changing the network will refresh the page. This might need refactoring anyway at some point, leaving it for now.

juhah commented on app/src/shared/web3/web3Poller/web3Poller.js:
```diff
@@ -0,0 +1,185 @@
+
+    pollWeb3 = () => (
+        this.fetchWeb3Account()
+            .then(this.startWeb3Poll, this.startWeb3Poll)
```
Yep, I tried `finally` first but the tests broke. Added the error logging now 👍

juhah commented:
@timoxley @mondoreale Please have another look when you have time, I've addressed the raised issues.

Also, I've refactored `createIdentity` service using `async`/`await` in #198 (https://github.com/streamr-dev/streamr-platform/blob/bf9988a44ff9aca63a08afd27dcd3539ece37b62/app/src/shared/modules/integrationKey/services.js#L43)
